### DISCUSSION TITLE
A Question about when should payment plugin set order status to on hold

### DISCUSSION THREAD

**Discussion by TeaDrinkerPlus** [woocommerce/woocommerce#50514] [Checkout Flow]
Hello, I am a payment plugin developer working on a plugin for digital wallet payments.

The payment process is as follows: when a user selects my plugin’s payment method at checkout, my plugin will request a QR code link to complete the wallet payment from the server. The link is then returned to the checkout page, allowing the user to proceed with the QR code payment.

After the user completes the scan, the wallet will send a notification to the WooCommerce merchant. My plugin listens for this notification and updates the order status to either successful or failed.

Currently, my design sets the order status to on-hold once the QR code link is successfully obtained. This is because the subsequent payment process requires user interaction to complete via the scan. Setting the status to on-hold also clears the cart to prevent users from choosing other payment plugins or methods, which could result in duplicate payments.

However, when reviewing other payment plugins, I noticed that many of them keep the order status as pending for such QR code payment scenarios.

This confuses me because an order in pending status can be subject to duplicate payments, and it does not clearly reflect that the user has selected another payment method and is waiting for payment completion.

Could you please advise if WooCommerce has any recommended handling for this logic?

**Comment by mikejolley** [comment]
I would say it depends if this process happens before or after pressing the "place order" button.

If the user has placed order and selected the QR code payment method, yes, using `on-hold` is correct since you're waiting for the user to complete an action for the selected gateway. As you said, this ensures the cart is cleared and the order data is in the system.

This matches the offline gateways like BACs where the user has selected that method, but the order is on-hold until they send the funds successfully.

**Comment by TeaDrinkerPlus** [comment]
Regarding this issue, I’d like to provide an update.
I discovered that WooCommerce, by default, sends a successful order email when an order is marked as "on-hold." If we set the status to "on-hold" for QR-code payments, this may cause confusion for customers.
